### ISSUE TITLE
Keyboard is not focused to value input when editing modulation values from context menu

### ISSUE THREAD

**Issue by xard-dev** [surge-synthesizer/surge#5758]
**Bug Description:**
Keyboard is not focused to value input when editing modulation values from context menu and the field needs to be clicked before a value can be written.

**Surge Version**
* Version: Surge XT 0.99.nightly.320cd70
* Plugin Type: VST3, LV2
* Bitness: 64-bit

**Reproduction Steps:**
1. Open control like Filter cutoff context menu
2. Click on any modulation **edit icons** OR click **Add modulation from** context menu item adding a new modulation target
3. Type value

**Expected Behavior:**
A value can be inserted directly.

**Screenshots:**
![image](https://user-images.githubusercontent.com/65303522/149224995-fc6945de-2e86-4b07-b0be-4d92b1d2cb3b.png)

**Computer Information (please complete the following information):**
 - OS: Ubuntu 20.04
 - Host: Carla, Ardour
 - Version: 2.1.0, 6.9

**Additional Information:**
When editing control value the keyboard is automatically focused to the value field.

**Comment by baconpaul** [comment]
OK this isn't a problem on mac but i happen to be in linux now so let me look

**Comment by baconpaul** [comment]
oh and confirmed that on linux it does happen. and it is annoying.

**Comment by baconpaul** [comment]
hmm i have no idea how to fix this
it's super annoying but i don't think i would block on it. I thought it might be our custom focus handlers but it isn't.

Oh @xard-dev I'm also interested if that , as . works properly with the parsing?

**Comment by xard-dev** [comment]
For my regional settings decimal separator comma , works as expected but period is ignored and the values after period are discarded.

**Comment by xard-dev** [comment]
And If I temporarily change to LC_ALL=C the period . is used instead of comma , as expected.

**Comment by baconpaul** [comment]
great thanks!

**Comment by j5v** [comment]
I came here to report this bug.
What might help:  While the 'edit value' box is being displayed, I can right-click again and again select the edit icon in the context menu. If I can get the value to focus, then repeated attempts will always be in focus. However, if the 'edit value' box is not focused, then it might, or might not focus next time.
I suspect a race condition between display and focus.

**Comment by baconpaul** [comment]
Yeah it’s all single threaded but I wonder if I need a show message to get a cycle before the focus message. Excellent observation

**Comment by SamiSieranoja** [comment]
I have the same issue. In my case, clicking the input box does not seem to help. Seems like the focus automatically leaves the input. If I click the input box and press a key at the same time, then I might be able to input one character, but not more. Tested in Reaper and Ardour. Tested with surge versions 1.3.4 and Nightly build 2025-12-11.  Ubuntu 24.04.3, under X11.

**Comment by xard-dev** [comment]
While there are some input focus issues under Xorg I can confirm that all the input issues are by order of magnitude worse under Wayland / xWayland which makes the keyboard input almost unusable currently. Though these are likely xWayland caused but still some investigation is required.

**Comment by xard-dev** [comment]
When it comes to the original context of this ticket I discovered that the focus transfer to the edit field actually DOES work in the LV2 version of Surge XT but not in the VST3 version.

**Comment by mkruselj** [comment]
So do you think it's still connected to Wayland somehow?

**Comment by xard-dev** [comment]
> So do you think it's still connected to Wayland somehow?

Sorry I brought up Wayland here as the ticket originally has been verified on Xorg and the main issue is still present on Xorg.

There are plenty of xWayland specific issues I cannot reproduce on Xorg, but I'm not quite sure if the Surge project would appreciate creating tickets about them here considering that everything should work like before under xWayland in ideal conditions. (Then again, I have no idea which would be a good place to start reporting these findings as finding correct places to bring up the findings to upstream is... complicated)

**Comment by mkruselj** [comment]
The biggest issue here IMO is that it's likely an interop issue between JUCE and Linux WMs... So it's not something that should be on us to fix, I feel...

**Comment by baconpaul** [comment]
My guess is the accessibility grab focus is screwing with something 

I need to try it on Linux. I may be able to make those grabs optional and off but I’m not sure 

If you shift menu there’s a developer menu which has a focus debugger - running with that could be useful

**Comment by xard-dev** [comment]
> If you shift menu there’s a developer menu which has a focus debugger - running with that could be useful

Thanks, I'll try that!

So far I've learned that the potential layers where input and focus problems can be are:
The plugin itself  > JUCE > xWayland > Mutter (under Gnome)

**Comment by xard-dev** [comment]
It looks like that **Shift + Menu** does not open the developer options but I looked the source code and the way to show the menu item is by opening it with secondary mouse button (typically right but not always).

The focus debugger draws a rectangle on the focused item and for me the focus jumps directly and incorrectly to the menu button most of the time I click any widgets with text input. (this is on xWayland / Fedora 43)

**Comment by mkruselj** [comment]
Right clicking the bottom right Menu button should show the Developer submenu there. Or shift+right click in empty parts of background.

**Comment by baconpaul** [comment]
yeah sorry two finger click or shift click are the trackpad gestures. should have said right click. my bad.

**Comment by xard-dev** [comment]
A little update here is that as I've been playing around with the Shortcircuit XT Beta I have no issues in that plugin with the mouse / keyboard input (even these value edit widgets work as expected) on Fedora 43 / Wayland.

**Comment by mkruselj** [comment]
@xard-dev Just curious - is the issue outlined in the first post still happening even with latest nightly? Just checking.

**Comment by mkruselj** [comment]
I unleashed Opus 5 at this issue:
___
I went through this against the JUCE we currently pin (8.0.12 plus our picks) to work out what is actually happening. Summary up front: this is **two distinct JUCE defects** that compose, both live in JUCE's X11 and focus code, and neither has been fixed upstream. Nothing here is Surge's own doing, but both are small enough that we could carry patches in our fork and push them upstream.

## Defect 1: the peer restores stale focus after a popup menu closes

Three pieces of JUCE interact.

**The menu item's action is deferred.** In `modules/juce_gui_basics/menus/juce_PopupMenu.cpp`, `MenuWindow::hide()` calls `exitModalState (resultID)` and then `MessageManager::callAsync (item->action)`. So everything Surge does in response to the click — `promptForUserValueEntry()` → `typeinParamEditor->grabFocus()` → `textEd->grabKeyboardFocus()` — runs one message-loop turn *after* the menu has already gone away.

**The peer caches and unconditionally restores the previous focus.** `ComponentPeer::handleFocusLoss()` (`modules/juce_gui_basics/windows/juce_ComponentPeer.cpp:381`) stashes `currentlyFocusedComponent` into `lastFocusedComponent` when the menu window takes native focus. When native focus comes back, `handleFocusGain()` (line 362) does:

```cpp
Component::currentlyFocusedComponent = lastFocusedComponent;
```

with no check for whether something else has taken focus in the meantime.

**On X11 that focus change is asynchronous.** `XWindowSystem::toFront` doesn't set focus directly, it `XSendEvent`s `_NET_ACTIVE_WINDOW` to the root window and lets the window manager act whenever it likes. Ordinary focus changes arrive as queued `FocusIn` events on the display FD and are dispatched whenever the event loop reaches them.

So the losing order is:

1. Right-click — the slider or the menu button holds JUCE keyboard focus.
2. The menu window (a separate override-redirect X window) takes native focus → `handleFocusLoss()` → `lastFocusedComponent = the slider`.
3. Item clicked → `exitModalState` → `PopupMenuCompletionCallback::modalStateFinished` destroys the menu window, calls `topLevel->toFront (true)` (an asynchronous request to the WM) and restores focus to the pre-menu component.
4. The deferred action runs. Surge shows the type-in and grabs focus on the text editor. Correct at this instant.
5. **Then** the queued `FocusIn` is dispatched → `handleFocusGain()` → `currentlyFocusedComponent = lastFocusedComponent` → focus snaps back to the slider or the menu button.

That is precisely what @xard-dev saw with the focus debugger ("the focus jumps directly and incorrectly to the menu button"), and it is @j5v's 2022 race intuition, which was right.

Whether you lose the race depends on how the X display FD and JUCE's internal message socketpair get interleaved, which is why the VST3/LV2 split @xard-dev found isn't mysterious: the LV2 wrapper pumps everything itself via `messageThread->processPendingEvents()`, polling all FDs in one pass in a fixed order, while the VST3 wrapper hands each FD to the host's `Steinberg::Linux::IRunLoop` and lets Ardour/Reaper/Carla decide the servicing order.

## Defect 2: JUCE acts on X11 pseudo focus events

This one came out of an unrelated codebase. The Usual Suspects hit it in Gearmulator and left a note in [`juceRmlComponent.cpp`](https://github.com/dsp56300/gearmulator/blob/112edafc6d3e1454ba0233d9c81a9dd391a7327c/source/framework/juce/juceRmlUi/juceRmlComponent.cpp#L827):

```cpp
// We skip the mouse leave on Linux because apparently focusLost is called when the mouse button
// is released?! https://tus.youtrack.cloud/tickets/BUG-10084/
#if JUCE_WINDOWS || JUCE_MAC
    m_rmlContext->ProcessMouseLeave();
#endif
```

`Component::focusLost` is only reached through `internalKeyboardFocusLoss`, which on Linux means `ComponentPeer::handleFocusLoss()` — the exact function that arms `lastFocusedComponent` above. So a second team, coming at this from a completely different direction, independently found that JUCE's X11 peer reports keyboard focus loss on plain mouse button release.

The cause is visible in the X event dispatch table in `modules/juce_gui_basics/native/juce_XWindowSystem_linux.cpp`:

```cpp
case ButtonRelease: handleButtonReleaseEvent (peer, event.xbutton);   break;
case LeaveNotify:   handleLeaveNotifyEvent   (peer, event.xcrossing); break;
case FocusIn:       handleFocusInEvent       (peer);                  break;
case FocusOut:      handleFocusOutEvent      (peer);                  break;
```

`FocusIn`/`FocusOut` are the only two handlers in the entire table that don't receive their event struct. The `XFocusChangeEvent` — and with it `mode` (`NotifyNormal` / `NotifyGrab` / `NotifyUngrab` / `NotifyWhileGrabbed`) and `detail` (`NotifyPointer`, `NotifyPointerRoot`, `NotifyInferior`, …) — is discarded before any logic runs.

X11 emits `FocusOut(NotifyGrab)` / `FocusIn(NotifyUngrab)` pairs around pointer and keyboard grabs. These are grab bookkeeping: the keyboard focus has not moved, and the conventional handling is to ignore them. JUCE treats every one as a real focus transfer.

JUCE already knows this pattern, which makes it a bit galling — twenty lines above, `handleLeaveNotifyEvent` does exactly the right thing:

```cpp
// Suppress the normal leave if we've got a pointer grab, or if
// it's a bogus one caused by clicking a mouse button when running
// in a Window manager
if (((! ...isAnyMouseButtonDown()) && leaveEvent.mode == NotifyNormal) || leaveEvent.mode == NotifyUngrab)
```

Grab-aware for crossing events, grab-blind for focus events.

## How the two defects split the reports

Defect 1 explains the original report and @susquetzal's: focus is lost once, at the moment the type-in appears, on the way out of a context menu.

Defect 2 explains @SamiSieranoja's, which never fitted the popup-menu story:

> clicking the input box does not seem to help. Seems like the focus automatically leaves the input. If I click the input box and press a key at the same time, then I might be able to input one character, but not more.

Under defect 2 that is exact rather than approximate. Mouse-down focuses the text field, mouse-**up** generates the pseudo `FocusOut`/`FocusIn` pair, `handleFocusGain()` restores `lastFocusedComponent` (the slider), and focus leaves the field the instant you release the button. A key pressed while the button is still held gets through; nothing after it does.

Defect 1 is the specific trigger, defect 2 is a general focus-churn source, and defect 2 is what makes the churn destructive rather than merely noisy.

## Why nobody reproduces this on Windows or macOS

Structurally, not by luck. `HWNDComponentPeer::setVisible` (`modules/juce_gui_basics/native/juce_Windowing_windows.cpp:1502`) shows every JUCE window with `SW_SHOWNA` — show *without* activating — and `PopupMenu::showWithOptionalCallback` uses `window->toFront (false)` and `enterModalState (false, ...)`, so no activation and no keyboard focus grab. A popup menu therefore never takes OS focus away from the plugin window, `WM_KILLFOCUS` never arrives, `handleFocusLoss()` never runs, and `lastFocusedComponent` is never armed. The steal needs that armed pointer, so there is nothing to fire. Defect 2 is X11-specific by construction.

I checked whether any of that was different in January 2022 and it wasn't: `SW_SHOWNA` predates the whole window, `PopupMenuCompletionCallback` lost its `prevFocused` fallback in JUCE commit `c2f661171a` (2021-10-19, three months before this thread got going), and `MessageManager::callAsync (item->action)` has been there since 2019.

That leaves one loose end. @j5v, you were on Windows 64-bit standalone when you confirmed this in January 2022, and the mechanism above can't produce what you saw. Your description also reads differently from the Linux reports — "if I can get the value to focus, then repeated attempts will always be in focus" is sticky, a one-time initialization failure, whereas the Linux bug re-steals on every invocation because `lastFocusedComponent` is re-armed each time the menu opens. The build in the original report (`320cd70`) is literally the same-day merge of "Focus Management after Typeins" (#5757), and Surge's type-in focus code churned for months afterwards, so there was very likely a separate Windows-side defect in that window that has since been fixed. There's also been a standalone initial-keyboard-focus bug in Surge right up until #8568 landed recently, which is at least a candidate. If you get a chance, could you retest on a current Windows nightly, both standalone and VST3? If it comes back clean we can treat the Windows half as closed and this issue becomes purely the X11 problem.

## Upstream status

Not fixed. I diffed our pin against JUCE `develop` (683 commits ahead, tip 2026-08-28, the JUCE 9 preview line). Nothing touches `handleFocusGain`, `PopupMenuCompletionCallback`, or X11 focus delivery. The Linux commits in that range are XInput2, XRandR, vblank timing, XEmbed mapping flags, and popup menu *accessibility*, all unrelated. Waiting for upstream is not a plan.

For the record, the accessibility hypothesis from earlier in the thread doesn't hold up — `AccessibilityHandler` isn't in this path at all.

## Candidate fixes, not yet applied

**For defect 1**, a guard in `ComponentPeer::handleFocusGain`:

```cpp
if (component.isParentOf (Component::currentlyFocusedComponent))
    return; // something grabbed focus while we were natively unfocused; don't clobber it
```

`isParentOf (nullptr)` returns false, and in the normal case `handleFocusLoss` has already nulled `currentlyFocusedComponent`, so the ordinary path is unchanged. It only differs in exactly the case we want to fix. Low risk.

**For defect 2**, pass the event through and filter on it:

```cpp
case FocusIn:   handleFocusInEvent  (peer, event.xfocus); break;
case FocusOut:  handleFocusOutEvent (peer, event.xfocus); break;

// NotifyGrab/NotifyUngrab are grab bookkeeping; the keyboard focus hasn't moved.
static bool isRealFocusChange (const XFocusChangeEvent& e)
{
    return (e.mode == NotifyNormal || e.mode == NotifyWhileGrabbed)
             && e.detail != NotifyPointer
             && e.detail != NotifyPointerRoot;
}
```

This one is riskier and wants real testing before it goes anywhere: it changes behavior for every JUCE window on Linux in the process — standalone, plugin, and the host's own embedding — and getting the `NotifyWhileGrabbed` or `NotifyInferior` cases wrong would mean windows stop noticing genuine focus changes, which is a worse failure than what we have now. It needs exercising under a reparenting WM and under xWayland, in more than one host.

**Comment by mkruselj** [comment]
Filed upstream as juce-framework/JUCE#1754, covering both defects with the suggested patches. Leaving this open to track it; if JUCE fixes it we get it for free on the next submodule bump, and if it stalls we can carry the picks in our fork.
